python/graphstorm/gconstruct/transform.py

```diff
@@ -935,6 +935,143 @@ def call(self, feats):
+    """ Translate input data into node ids for hard negative stored as edge features
+
+    Parameters
+    ----------
+    col_name : str
+        The name of the column that contains the feature.
+    feat_name : str
+        The feature name used in the constructed graph.
+    separator : str
+        The separator to split data into multiple node ids.
+    """
+    def __init__(self, col_name, feat_name, separator=None):
+        self._target_ntype = None
+        self._target_etype = None
+        self._nid_map = None
+        self._separator = separator
+        super().__init__(col_name, feat_name, out_dtype=np.int64)
+
+    def set_target_etype(self, etype):
+        """ Set the etype of this hard edge negative transformation ops
+            and associated hard negative information. For example,
+            self._target_ntype.
+
+        Parameters
+        ----------
+        etype : tuple of str
+            The edge type the hard negatives belonging to.
+        """
+        raise NotImplementedError
+
+    @property
+    def target_etype(self):
+        """ The the edge type of this hard negative transformation.
+        """
+        return self._target_etype
+
+    @property
+    def neg_ntype(self):
+        """ Return the node type of hard negatives
+        """
+        return self._target_ntype
+
+    def set_id_maps(self, id_maps):
+        """ Set ID mapping for converting raw string ID to Graph ID
+        """
+        assert self._target_ntype is not None, \
+            "The target node type should be set, it can be the source node type " \
+            "or the destination node type depending on the hard negative case."
+        assert self._target_ntype in id_maps, \
+            f"The nid mapping should have the mapping for {self._target_ntype}. " \
+            f"But only has {id_maps.keys()}"
+        self._nid_map = id_maps
+
+    def pre_process(self, feats):
+        """ Pre-process data
+
+            Not all the edges have the same number of hard negatives.
+            Thus we need to know the maxmun number of hard negatives first.
+
+        Parameters
+        ----------
+        feats:
+            feats to be processed
+        """
+        assert isinstance(feats, (np.ndarray, ExtMemArrayWrapper)), \
+            "Feature of HardEdgeNegativeTransform must be numpy array or ExtMemArray"
+
+        if self._separator is None:
+            max_dim = feats.shape[1]
+        else:
+            assert len(feats.shape) == 1 or feats.shape[1] == 1, \
+                "When a separator is given, the input feats must be a list of strings."
+
+
+            if feats.dtype.type is not np.str_:
+                logging.warning("When a separator is given, "
+                                "the input feats must be a list of strings."
+                                "But we get %s. Will convert the input feats into str array",
+                                feats.dtype.type)
+            feats = feats.astype(str)
```

Comment:
what is `str`?

Reply:
We are converting feats to string numpy array. (When loading data from Parquet, sometimes, the numpy array dtype is `object`.